client/can.go

```diff
@@ -0,0 +1,213 @@
+		log.Println(errors.Wrap(err, "CanBusClient: Error creating Socket object"))
+	}
+
+	canMsgRx := make(chan canbus.Frame)
+
+	closePort := func() {
+		socket.Close()
+	}
+
+	listener := func() {
+		for {
+			frame, err := socket.Recv()
+			if err != nil {
+				log.Println(errors.Wrap(err, "CanBusClient: Error recieving CAN frame"))
+			}
+			canMsgRx <- frame
+		}
+	}
+
+	openPort := func() {
+
+		iface, err := net.InterfaceByName(cb.config.Interface)
+		_ = iface
+		if err != nil {
+			log.Println(errors.Wrap(err, "CanBusClient: CAN interface not found"))
+		}
+
+		// TODO: figure out how to handle interface
+		/*
+		//if iface.Flags&net.FlagUp == 0 {
+		// bring up CAN interface
+		err = exec.Command("ip", "link", "set", cb.config.Interface, "type",
+			"can", "bitrate", cb.config.BusSpeed).Run()
+		log.Println("Bringing up IP link with:", cb.config.Interface, cb.config.BusSpeed)
+		if err != nil {
+			log.Println(errors.Wrap(err, "Error configuring internal CAN interface"))
+		}
+
+		err = exec.Command("ip", "link", "set", cb.config.Interface, "up").Run()
+		if err != nil {
+			log.Println(errors.Wrap(err, "Error bringing up internal can interface"))
+		}
+		*/
+		/*
+			} else {
+				// Handle case where interface is already up and bus speed may be wrong
+				log.Println("Error bringing up internal CAN interface, already set up.")
+			}
+		*/
+		// TODO: figure out a way to set accepted CAN id's for filtering and decoding
+		/*
+		var filters []unix.CanFilter
+		for _, id := range cb.config.CanIdsAccepted {
+			filters = append(filters, unix.CanFilter{Id: id, Mask: unix.CAN_EFF_MASK})
+			log.Println("CanBusClient: set filter {Id: %X, Mask: %X}", id, unix.CAN_EFF_MASK)
+		}
+		socket.SetFilters(filters[:])
+		*/
+
+		err = socket.Bind(cb.config.Interface)
+		if err != nil {
+			log.Println(errors.Wrap(err, "Error binding to CAN interface"))
+		}
+		go listener()
+	}
+
+	openPort()
+
+	for {
+		select {
+		case <-cb.stop:
+			log.Println("CanBusClient: stopping CAN bus client: ", cb.config.Description)
+			closePort()
+			return nil
+
+		case frame := <-canMsgRx:
+
+			log.Printf("CanBusClient: got %X, data length: %v\n", frame.ID, len(frame.Data))
+
+			points := make(data.Points, 2)
+
+			// FIXME decode data based on information in config
+			points[0].Time = time.Now()
+			points[1].Time = time.Now()
+			points[0].Key = "FE48-1862-WheelBasedSpeed"
+			points[1].Key = "FE48-1864-WheelBasedDirection"
+			points[0].Value = 0
+			points[1].Value = 0
+
+			// Send the points
+			if len(points) > 0 {
+				err = SendPoints(cb.nc, cb.natsSub, points, false)
+				if err != nil {
+					log.Println(errors.Wrap(err, "CanBusClient: error sending points received from CAN bus: "))
+				} else {
+					log.Println("CanBusClient: successfully sent points")
+				}
+			}
+
+		case pts := <-cb.newPoints:
+			for _, p := range pts.Points {
+				if p.Type == data.PointTypePort ||
+					p.Type == data.PointTypeBaud ||
+					p.Type == data.PointTypeDisable {
```

Comment:
you've probably not got to this yet, but his is a good place to close and reopen the port if you change settings

Reply:
Good to know!